### PR TITLE
Retro Points Fix (Revisions, Bonus, Hardcore, UnPopular games)

### PR DESCRIPTION
**Pull request by Xymjak**
Because the post still is linked, 2023 update for context:

1. I was told that I'm not allowed to change the structure or db. Most of this code is workaround against the DB then (IDK about now) not having `SELECT ID, NumPlayers FROM Revisions WHERE GameId = $gameId`

2. I was told that I'm not allowed to change the existing formula so the code is just "bonus" on top of the old formula.

3. And after that I was told that not allowed to change stuff until V2...

The original message:

Retropoints used to have these complains for years:

1. Revisions are inadequately increase the points

5. Bonus sets have low amounts of retropoints despite their difficulty

6. Ultra hard achievements with 0 unlocks don't show retropoints

7. Unpopular hard games have lower retropoints than easier popular games

8. Retropoins use softcore instead of hardcore

So here is the changes without updating the DB - because making a gigantic update to the DB when it will be updated in V2 is pretty useless, and this code will be easy to adapt to V2 with few small changes.

1. Revisions will use players that were started to play only after a revision. Well... no quite. At first I did this naive approach and immediately noticed that the retro-ratio becomes below 1 - because old players immediately unlocks new achievements but they started to play before a revision. So I added small calculated boost to players (checking the exact number of old players is too taxing without DB-modifications) Without this boost: for 3 months points are below what they should. With boost: for 1-2 weeks points are a little above (50 times less than they were before) and after they are normal.

Tests (with 4 months old database, so numbers are a little different from the site):

| Game | Before | After | Change | Comment |
|----------|----------|---------|------------|--------|
| SOR3 | 15000 | 2600 | -83% | New Rev |
| Final Fight 2 | 3600 | 1100 | -69% | 
| R-Type | 23000 | 8000 | -65% | Guess this is a hard set after all  |
| SOR | 7000 | 2400 | -66% | 
| Darkwing Duck | 4100 | 1500 | -63% | 
| NG Shadow | 2400 | 1000 | -58% | 
| Clock Tower | 1400 | 650 | -54% | 
| SMB2J | 7500 | 4300 | -43% | 2018 Rev |
| SMB | 6200 | 3600 | -42% | Old Rev |
| SMW | 3000 | 1750 | -42% | Old Rev |
| Contra GB | 2700 | 1600 | -41% | 
| Castlevania | 2000 | 1300 | -35% | 2017 rev |
| River City | 2400 | 1600 | -33% |
| DKC | 1100 | 800 | -27% | 2016 rev, the DB is before 2020 rev |
| MMX | 2000 | 1500 | -25% |  2017 rev 

2. Bonus sets are counts as revisions for the main set and calculated the same way ie using only players from the main set after their creation or revision... except they are using only 25% of players. Why? Because otherwise they fly off into outer space (just check FF7 in the table below).
5. Achievement with 0 unlocks shows retropoints.

| Game | Before | After | Change | Comment |
|----------|----------|---------|------------|--------|
| SM64 Bonus | 1300 | 5300 | 308% | 
| Little MM Bonus | 250 | 1250 | 400% | 
| Castlevania Bonus | 1400 | 8666 | 519% | 
| MMX Bonus | 500 | 3300 | 560% | 
| Mega Man Bonus | 600 | 7000 | 1070% |
| Akumajou Sp Bonus | 500 | 5900 | 1080% | 
| FF7 Bonus | 2000 | 44000 | 2100% | Over 25k are from 0 unlocks achievements |

6. Added small bonus to unpopular games. IT'S NOT MULTIPLIER, the boost depends on game's difficulty

| Game | Before | After | Change | Comment |
|----------|----------|---------|------------|--------|
| Daedalian Opus | 530 | 550 | 4% | Easy game so the boost is really small |
| Croc | 170 | 190 | 12% | 
| Frankenstein | 1000 | 1200 | 20% | Only few hard achievements |
| Densetsu no Stafy | 900 | 1100 | 22% | Only few hard achievements |
| Kaizou Choujin | 520 | 660 | 27% |
| Magical Taruruuto | 540 | 720 | 33% |
| Sylvan Tale | 1000 | 1850 | 85% | A lot of hard achievements |

7. And here is the changes from excluding softcore players. 

| Game | Before | After | Change | Comment |
|----------|----------|---------|------------|----|
| Sonic 1 | 1100 | 800 | -27% | 
| Wario Land 4 | 1400 | 1200 | -14% | 
| Metroid | 1000 | 900 | -10% | 
| Gaia | 2300 | 2100 | -9% | 
| Goof Troop | 1200 | 1100 | -8% | 
| KDL2 | 1700 | 1600 | -6%
| Mickey to Donald 3 | 930 | 920 | -1% | 
| DK | 880 | 880 | 0% | 
| MM6 | 750 | 750 | 0% | 
| Ninja Gaiden 2 | 1500 | 1500 | 0% | 
| CV: Bloodlines | 4200 | 4300 | 2% | 
| Valis | 200 | 210 | 5% | 
| Doom GBA | 600 | 630 | 5% | 
| Faxanadu | 950 | 1000 | 5% | 
| Mario Bros | 1900 | 2100 | 11% | 
| DK3 | 2700 | 4000 | 48% | 33% of "masteries" for this game are "softcore" so this game is increased a lot |

### CODE REVIEW

**Comment by Blazekickn** [comment]
lets fucking gooo

File: lib/database/points.php [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+        if (!$gotParent) {
+            $realID = $gameID;
+        }
+
+        return TrueRetroRatio_WorkAroundV1::
+        calculatePlayersPerRevision($achsGrouped, $realID, $gotParent, $HC, $checkUnTrThr);
+    }
+}
+
+class TrueRetroRatio
```
Comment (MagmarFire) [inline lib/database/points.php:329]: All classes need to be in their own file.

File: lib/database/points.php [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+    }
+
+    private function getPercentageOfBoost($players)
+    {
+        $boost0At = 1000;
+        if ($players >= $boost0At) {
+            return 0;
+        }
+
+        //$boost50At = 200;
```
Comment (MagmarFire) [inline lib/database/points.php:370]: If this is code that ended up becoming unused, it should be removed.

**Review by MagmarFire — CHANGES_REQUESTED** [review]

File: lib/database/points.php [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+
+function recalculateTrueRatio($gameID)
+{
+    sanitize_sql_inputs($gameID);
+
+    return (new TrueRetroRatio())->Recalculate($gameID);
+
+    /*
+     * PREVIOUS VERSION
+     */
```
[the remark this answers is not in the document]
Comment (luchaos) [reply lib/database/points.php:11]: Converted to draft. I think i added some of the commented code as reference whilst trying to make it work again. I'll take care of it.

File: lib/database/points.php [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+    $query = $query . " WHERE gd.ID = $gameID AND aw.HardcoreMode = $hardcore AND ach.Flags = 3";
+
+    if ($checkUnTracked) {
+        $query = $query . " AND NOT ua.Untracked";
+    }
+    $data = getDbData($query);
+    return ($data == null) ? null : $data[0]->UniquePlayers;
+}
+
+class TrueRetroRatio_WorkAroundV1
```
[the remark this answers is not in the document]
Comment (Xymjak) [reply lib/database/points.php:117]: Originally I was told that I'm not allowed to do that. (Or even not allowed to use OOP.)

File: lib/database/points.php (hunk 2 above)
Comment (Xymjak) [reply lib/database/points.php]: It not as much as code, as explanation how the "magic number" few lines below appeared.

**Comment by wescopeland** [comment]
What is the next action we can take on this PR, if any?

**Comment by wescopeland** [comment]
Thank you for your contribution to RAWeb.

We've noticed that this pull request has been in draft status for an extended period of time and the branch has now become stale due to significant changes in the codebase since the PR was originally opened. As it stands, reviewing this PR against the current version of the code is challenging, and unfortunately, we are unable to proceed with it in its current state.

We greatly appreciate your efforts and interest in improving RAWeb. For now, we will be closing this PR. However, we warmly invite you to update and reopen it when it aligns with the latest commit in `master` and you feel it is ready for review. Your contributions are always welcome.

Thanks for your understanding.

**Pull request closed without merging**